Crash on code mixing instances and type typed variables

<!--
  Use this form only if mypy reports an "INTERNAL ERROR" and/or gives a traceback.
  Please include the traceback and all other messages below (use `mypy --show-traceback`).
-->

**Crash Report**

INTERNAL ERROR

**Traceback**

```python-traceback
mwe-mypy2.py:19: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 1.20.0+dev.56c33f251cb6b3791297fdaa57453d4afaa8c4dc
Traceback (most recent call last):
  File "/home/.../.venv/bin/mypy", line 8, in <module>
    sys.exit(console_entry())
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/__main__.py", line 15, in console_entry
    main()
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/main.py", line 135, in main
    res, messages, blockers = run_build(sources, options, fscache, t0, stdout, stderr)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/main.py", line 219, in run_build
    res = build.build(sources, options, None, flush_errors, fscache, stdout, stderr)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/build.py", line 320, in build
    result = build_inner(
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/build.py", line 417, in build_inner
    graph = dispatch(sources, manager, stdout)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/build.py", line 3178, in dispatch
    process_graph(graph, manager)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/build.py", line 3588, in process_graph
    done, still_working, results = manager.wait_for_done(graph)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/build.py", line 1070, in wait_for_done
    process_stale_scc(graph, next_scc, self)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/build.py", line 3752, in process_stale_scc
    graph[id].type_check_first_pass()
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/build.py", line 2543, in type_check_first_pass
    self.type_checker().check_first_pass()
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checker.py", line 530, in check_first_pass
    self.accept(d)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checker.py", line 641, in accept
    stmt.accept(self)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/nodes.py", line 1004, in accept
    return visitor.visit_func_def(self)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checker.py", line 1198, in visit_func_def
    self._visit_func_def(defn)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checker.py", line 1202, in _visit_func_def
    self.check_func_item(defn, name=defn.name)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checker.py", line 1237, in check_func_item
    self.check_func_def(defn, typ, name, allow_empty)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checker.py", line 1510, in check_func_def
    self.accept(item.body)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checker.py", line 641, in accept
    stmt.accept(self)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/nodes.py", line 1582, in accept
    return visitor.visit_block(self)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checker.py", line 3159, in visit_block
    self.accept(s)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checker.py", line 641, in accept
    stmt.accept(self)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/nodes.py", line 1669, in accept
    return visitor.visit_assignment_stmt(self)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checker.py", line 3212, in visit_assignment_stmt
    self.check_assignment(s.lvalues[-1], s.rvalue, s.type is None, s.new_syntax)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checker.py", line 3387, in check_assignment
    rvalue_type, lvalue_type = self.check_simple_assignment(
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checker.py", line 4668, in check_simple_assignment
    rvalue_type = self.expr_checker.accept(
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checkexpr.py", line 6081, in accept
    typ = self.accept_maybe_cache(node, type_context=type_context)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checkexpr.py", line 6116, in accept_maybe_cache
    typ = node.accept(self)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/nodes.py", line 2298, in accept
    return visitor.visit_call_expr(self)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checkexpr.py", line 508, in visit_call_expr
    return self.visit_call_expr_inner(e, allow_none_return=allow_none_return)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checkexpr.py", line 645, in visit_call_expr_inner
    ret_type = self.check_call_expr_with_callee_type(
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checkexpr.py", line 1495, in check_call_expr_with_callee_type
    ret_type, callee_type = self.check_call(
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checkexpr.py", line 1588, in check_call
    return self.check_callable_call(
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checkexpr.py", line 1799, in check_callable_call
    arg_types = self.infer_arg_types_in_context(callee, args, arg_kinds, formal_to_actual)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checkexpr.py", line 1973, in infer_arg_types_in_context
    res[ai] = self.accept(args[ai], arg_type)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checkexpr.py", line 6083, in accept
    typ = node.accept(self)  # r-value type, when interpreted as a value expression
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/nodes.py", line 2198, in accept
    return visitor.visit_name_expr(self)
           ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checkexpr.py", line 357, in visit_name_expr
    narrowed = self.narrow_type_from_binder(e, result)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/checkexpr.py", line 6394, in narrow_type_from_binder
    narrowed = narrow_declared_type(known_type, restriction)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/meet.py", line 141, in narrow_declared_type
    narrow_declared_type(d, n)
    ~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/meet.py", line 222, in narrow_declared_type
    return meet_types(original_declared, original_narrowed)
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/meet.py", line 113, in meet_types
    return t.accept(TypeMeetVisitor(s))
           ~~~~~~~~^^^^^^^^^^^^^^^^^^^^
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/types.py", line 2846, in accept
    return visitor.visit_tuple_type(self)
           ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/meet.py", line 1079, in visit_tuple_type
    return t.copy_modified(items=[meet_types(it, self.s.args[0]) for it in t.items])
                                  ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^mwe-mypy2.py:19: : note: use --pdb to drop into pdb

  File "/home/.../.venv/lib/python3.13/site-packages/mypy/meet.py", line 113, in meet_types
    return t.accept(TypeMeetVisitor(s))
           ~~~~~~~~^^^^^^^^^^^^^^^^^^^^
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/types.py", line 3416, in accept
    return visitor.visit_union_type(self)
           ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/meet.py", line 793, in visit_union_type
    meets = [meet_types(x, self.s) for x in t.items]
             ~~~~~~~~~~^^^^^^^^^^^
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/meet.py", line 113, in meet_types
    return t.accept(TypeMeetVisitor(s))
           ~~~~~~~~^^^^^^^^^^^^^^^^^^^^
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/types.py", line 1250, in accept
    return visitor.visit_unpack_type(self)
           ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/home/.../.venv/lib/python3.13/site-packages/mypy/meet.py", line 845, in visit_unpack_type
    raise NotImplementedError
NotImplementedError: 
```

**To Reproduce**

```python
from typing import Hashable, cast

type Key[K: Hashable] = K | type[K]
type TP[K: Hashable] = tuple[Key[K], ...]
type Value[V] = V | type[V] | None
type T[K: Hashable, V] = tuple[*TP[K], Value[V]]

def f[K, V](one: TP[K] | T[K, V], other: TP[K] | T[K, V]) -> TP[K] | T[K, V] | None:
    if len(one) < len(other):
        return None
    if one[:len(other)] != other:
        return None
    return cast(TP[K] | T[K, V], one[len(other):])

def g[K, V](one: TP[K] | T[K, V], other: TP[K] | T[K, V]) -> int | None:
    d: TP[K] | T[K, V] | None = f(one, other)
    if d is None:
        return None
    n: int = len(d)
    return n
```

and then run `mypy` on it.

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: mypy 1.20.0+dev.56c33f251cb6b3791297fdaa57453d4afaa8c4dc (compiled: no)
- Mypy command-line flags: `--show-traceback`
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: Python 3.13.5
- Operating system and version: 6.12.57+deb13-amd64 `#1` SMP PREEMPT_DYNAMIC Debian 6.12.57-1 (2025-11-05) x86_64 GNU/Linux

<!--
You can freely edit this text, please remove all the lines
you believe are unnecessary.
-->

--- a/mypy/meet.py
+++ b/mypy/meet.py
@@ -1090,7 +1090,28 @@ class TypeMeetVisitor(TypeVisitor[ProperType]):
         elif isinstance(self.s, Instance):
             # meet(Tuple[t1, t2, <...>], Tuple[s, ...]) == Tuple[meet(t1, s), meet(t2, s), <...>].
             if self.s.type.fullname in TUPLE_LIKE_INSTANCE_NAMES and self.s.args:
-                return t.copy_modified(items=[meet_types(it, self.s.args[0]) for it in t.items])
+                arg = self.s.args[0]
+                new_items: list[Type] = []
+                for it in t.items:
+                    # Unpack items need to be handled by the caller.
+                    if isinstance(it, UnpackType):
+                        unpacked = get_proper_type(it.type)
+                        if isinstance(unpacked, TypeVarTupleType):
+                            # We can't infer anything in this case.
+                            new_arg = UninhabitedType()
+                            instance = unpacked.tuple_fallback
+                        else:
+                            assert (
+                                isinstance(unpacked, Instance)
+                                and unpacked.type.fullname == "builtins.tuple"
+                            )
+                            new_arg = meet_types(unpacked.args[0], arg)
+                            instance = unpacked
+                        new_items.append(UnpackType(instance.copy_modified(args=[new_arg])))
+                    else:
+                        # All other items can be processed in a regular way.
+                        new_items.append(meet_types(it, arg))
+                return t.copy_modified(items=new_items)
             elif is_proper_subtype(t, self.s):
                 # A named tuple that inherits from a normal class
                 return t

--- a/test-data/unit/check-typeis.test
+++ b/test-data/unit/check-typeis.test
@@ -213,8 +213,7 @@ def test5(t: tuple[A | B, ...]) -> None:
 
 def test6(t: tuple[B, Unpack[tuple[A | B, ...]], B]) -> None:
     if is_tuple_of_B(t):
-        # Should this be tuple[B, *tuple[B, ...], B]
-        reveal_type(t)  # N: Revealed type is "tuple[__main__.B, Never, __main__.B]"
+        reveal_type(t)  # N: Revealed type is "tuple[__main__.B, Unpack[builtins.tuple[__main__.B, ...]], __main__.B]"
     else:
         reveal_type(t)  # N: Revealed type is "tuple[__main__.B, Unpack[builtins.tuple[__main__.A | __main__.B, ...]], __main__.B]"
 [builtins fixtures/tuple.pyi]

--- a/test-data/unit/check-typevar-tuple.test
+++ b/test-data/unit/check-typevar-tuple.test
@@ -2873,3 +2873,15 @@ reveal_type(x)  # N: Revealed type is "builtins.list[tuple[()]]"
 reveal_type(y)  # N: Revealed type is "builtins.list[tuple[()]]"
 reveal_type(z)  # N: Revealed type is "builtins.list[tuple[()]]"
 [builtins fixtures/tuple.pyi]
+
+[case testMeetTupleWithUnpackVsInstanceWorks_no_verbose_reveal]
+from typing import TypeVar, Unpack, Union
+
+K = TypeVar("K")
+Prefix = tuple[Union[K, list[int]], ...]
+
+def g(x: K) -> None:
+    d: Union[Prefix[K], tuple[Unpack[Prefix[int]], int], None] = None
+    assert d is not None
+    reveal_type(d)  # N: Revealed type is "tuple[K | list[int], ...] | tuple[*tuple[int | list[int], ...], int]"
+[builtins fixtures/tuple.pyi]

